In [1]:
import numpy as np
import pandas as pd


In [2]:
df = pd.read_csv("creditcard.csv")

In [3]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
df.shape

(284807, 31)

In [5]:
list_of_columns = list(df.columns)
list_of_columns.remove("Time")
list_of_columns.remove("Amount")
list_of_columns.remove("Class")

In [6]:
list_of_df = []
for column in list_of_columns:
    array = np.array([df[column].value_counts(normalize=True).index.to_numpy(), 
                      df[column].value_counts(normalize=True).values])
    array = array.transpose()
    name = f"df_{column}"
    name = pd.DataFrame(array, columns=pd.MultiIndex.from_arrays(np.array([[column, column],["Values", "Freq"]])))
    list_of_df.append(name)

In [7]:
df_counts = pd.DataFrame()
for temp_df in list_of_df:
    df_counts = pd.concat([df_counts, temp_df], axis=1)

In [8]:
pd.options.display.max_columns = 100

In [9]:
df_counts

V1                  V2                  V3                  V4  \
          Values      Freq    Values      Freq    Values      Freq    Values   
0       2.055797  0.000270  0.166975  0.000270 -2.752041  0.000270 -0.842316   
1       1.245674  0.000270 -0.326668  0.000270  0.488306  0.000270  0.635322   
2       2.053311  0.000218  0.089735  0.000218 -1.681836  0.000218  0.454212   
3       1.302378  0.000211 -0.606529  0.000211 -0.681986  0.000211 -1.904603   
4       2.040211  0.000186 -0.146975  0.000186 -2.955934  0.000186 -0.578356   
...          ...       ...       ...       ...       ...       ...       ...   
275658 -0.069327  0.000004  0.520114  0.000004 -2.773663  0.000004  2.834682   
275659 -1.619420  0.000004 -0.419199  0.000004 -2.148874  0.000004  0.564523   
275660  1.174923  0.000004 -0.303794  0.000004  0.741123  0.000004 -1.193373   
275661  2.157306  0.000004  0.167497  0.000004 -2.696536  0.000004 -0.711142   
275662  2.287784  0.000004 -0.166213  0.000004  0.315988  0.000004  1.337592   

                        V5                  V6                  V7            \
            Freq    Values      Freq    Values      Freq    Values      Freq   
0       0.000270  2.463072  0.000270 -1.011073  0.000270 -0.432126  0.000270   
1       0.000270 -0.562777  0.000270  3.173856  0.000270  0.014953  0.000270   
2       0.000218  0.298310  0.000218 -0.953526  0.000218  0.152003  0.000218   
3       0.000211  1.326623  0.000211  3.436312  0.000211 -1.145127  0.000211   
4       0.000186  2.609358  0.000186  3.142642  0.000186 -0.416883  0.000186   
...          ...       ...       ...       ...       ...       ...       ...   
275658  0.000004 -0.448828  0.000004  0.923729  0.000004 -0.006237  0.000004   
275659  0.000004  0.250671  0.000004 -0.616964  0.000004  0.697378  0.000004   
275660  0.000004 -0.522516  0.000004 -1.010584  0.000004  1.130880  0.000004   
275661  0.000004 -0.649162  0.000004 -1.296312  0.000004  0.573042  0.000004   
275662  0.000004  0.245159  0.000004  0.880804  0.000004  0.223989  0.000004   

              V8                  V9                 V10                 V11  \
          Values      Freq    Values      Freq    Values      Freq    Values   
0      -0.160211  0.000270  0.170362  0.000270 -0.044575  0.000270 -0.356749   
1       0.727706  0.000270  0.608606  0.000270 -0.075186  0.000270  0.063504   
2      -0.207071  0.000218  0.587335  0.000218 -0.362047  0.000218 -0.589598   
3       0.959147  0.000211  1.671302  0.000211 -1.022946  0.000211 -0.191423   
4       0.784393  0.000186  0.359902  0.000186 -0.351075  0.000186  0.329651   
...          ...       ...       ...       ...       ...       ...       ...   
275658  0.294229  0.000004  0.340422  0.000004 -0.936097  0.000004  1.041802   
275659 -0.175237  0.000004  1.963203  0.000004 -1.625618  0.000004 -0.685356   
275660  1.829112  0.000004 -0.763159  0.000004  0.346382  0.000004 -0.028580   
275661  0.027999  0.000004  1.344202  0.000004  0.059576  0.000004 -1.010640   
275662  0.407707  0.000004 -0.562424  0.000004 -0.500907  0.000004  0.967952   

                       V12                 V13                 V14            \
            Freq    Values      Freq    Values      Freq    Values      Freq   
0       0.000270  0.350564  0.000270 -0.141238  0.000270  0.406969  0.000270   
1       0.000270 -0.073460  0.000270 -0.517760  0.000270  0.690972  0.000270   
2       0.000218 -0.174712  0.000218 -0.621127  0.000218 -0.703513  0.000218   
3       0.000211  0.631027  0.000211  0.031907  0.000211 -0.031425  0.000211   
4       0.000186  0.183508  0.000186 -0.272919  0.000186 -0.597437  0.000186   
...          ...       ...       ...       ...       ...       ...       ...   
275658  0.000004  1.026525  0.000004 -0.958312  0.000004  2.217835  0.000004   
275659  0.000004 -0.331234  0.000004  0.973101  0.000004  0.087633  0.000004   
275660  0.000004 -0.046139  0.000004  1.076993  0.000004 -0.383463  0.000004   
275661  0.0

#### либо я опять что-то не понял в предподготовке данных)))

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
Time      284807 non-null float64
V1        284807 non-null float64
V2        284807 non-null float64
V3        284807 non-null float64
V4        284807 non-null float64
V5        284807 non-null float64
V6        284807 non-null float64
V7        284807 non-null float64
V8        284807 non-null float64
V9        284807 non-null float64
V10       284807 non-null float64
V11       284807 non-null float64
V12       284807 non-null float64
V13       284807 non-null float64
V14       284807 non-null float64
V15       284807 non-null float64
V16       284807 non-null float64
V17       284807 non-null float64
V18       284807 non-null float64
V19       284807 non-null float64
V20       284807 non-null float64
V21       284807 non-null float64
V22       284807 non-null float64
V23       284807 non-null float64
V24       284807 non-null float64
V25       284807 non-null float64
V26  

#### пропущенных данных нет, и все содержат числовой тип данных

In [11]:
df.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   
5  0.260314 -0.568671 -0.371407  1.341262  0.359894 -0.358091 -0.137134   
6  0.081213  0.464960 -0.099254 -1.416907 -0.153826 -0.751063  0.167372   
7 -3.807864  0.615375  1.249376 -0.619468  0.291474  1.757964 -1.323865   
8  0.851084 -0.392048 -0.410430 -0.705117 -0.110452 -0.286254  0.074355   
9  0.069539 -0.736727 -0.366846  1.017614  0.836390  1.006844 -0.443523   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   
5  0.517617  0.401726 -0.058133  0.068653 -0.033194  0.084968 -0.208254   
6  0.050144 -0.443587  0.002821 -0.611987 -0.045575 -0.219633 -0.167716   
7  0.686133 -0.076127 -1.222127 -0.358222  0.324505 -0.156742  1.943465   
8 -0.328783 -0.210077 -0.499768  0.118765  0.570328  0.052736 -0.073425   
9  0.150219  0.739453 -0.540980  0.476677  0.451773  0.203711 -0.246914   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
5 -0.559825 -0.026398 -0.371427 -0.232794  0.105915  0.253844  0.081080   
6 -0.270710 -0.154104 -0.780055  0.750137 -0.257237  0.034507  0.005168   
7 -1.015455  0.057504 -0.649709 -0.415267 -0.051634 -1.206921 -1.085339   
8 -0.268092 -0.204233  1.011592  0.373205 -0.384157  0.011747  0.142404   
9 -0.633753 -0.120794 -0.385050 -0.069733  0.094199  0.246219  0.083076   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0  
5    3.67      0  
6    4.99      0  
7   40.80      0  
8   93.20      0  
9    3.68      0

In [20]:
X = df.loc[:, :"Amount"].copy()

In [21]:
y = pd.Series(df["Class"].copy())

In [38]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100, stratify=y)

In [35]:
print(f" X_train {X_train.shape} \n X_test {X_test.shape}  \n y_train {y_train.shape} \n y_test {y_test.shape}" )

 X_train (199364, 30) 
 X_test (85443, 30)  
 y_train (199364,) 
 y_test (85443,)


In [41]:
parametrs = [{
    "n_estimators": [10, 15],
    "max_features": np.arange(3, 5),
    "max_depth": np.arange(4,7)
}]

In [42]:
clf = GridSearchCV(estimator=RandomForestClassifier(random_state=100),
                   param_grid=parametrs,
                   scoring="roc_auc",
                   cv=3)

In [43]:
clf.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise-deprecating',
             estimator=RandomForestClassifier(bootstrap=True, class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators='warn', n_jobs=None,
                                              oob_score=False, random_state=100,
                                              verbose=0, warm_start=False),
             iid='warn', n_jobs=None,
             param_grid=[

In [44]:
clf.best_params_

{'max_depth': 6, 'max_features': 3, 'n_estimators': 15}

In [49]:
y_pred_proba = clf.predict_proba(X_test)

In [56]:
y_pred_proba = y_pred_proba.transpose()[1]

In [57]:
y_pred_proba

array([0.00092917, 0.00029521, 0.00028215, ..., 0.00028215, 0.0006822 ,
       0.01246098])

In [61]:
y_test = y_test.to_numpy()

In [62]:
from sklearn.metrics import roc_auc_score

In [66]:
AUC = roc_auc_score(y_test, y_pred_proba)

In [67]:
AUC

0.9462664156037156

AUC на уровне 0.9462664156037156, говорит о том, что полученная модель получилась с хорошей предсказательной силой и корректно определяет True Positive и False Negative